selftest.py

```diff
@@ -82,11 +138,52 @@ def run_self_tests(cls, _db, constructor_method=None, *args, **kwargs):
             result.success = False
         finally:
             result.end = datetime.datetime.utcnow()
-        yield result
+        results.append(result)
 
         if instance:
             for result in instance._run_self_tests(_db):
-                yield result
+                results.append(result)
+            integration = instance.external_integration(_db)
+        end = datetime.datetime.utcnow()
+
+        # Format the results in a useful way.
+        value = dict(
+            start=AtomFeed._strftime(start),
+            end=AtomFeed._strftime(end),
+            duration = (end-start).total_seconds(),
+            results = [x.to_dict for x in results]
+        )
+
+        # Store the formatted results in the database, if we can find
+        # a place to store them.
+        if integration:
+            integration.setting(
+                cls.SELF_TEST_RESULTS_SETTING
+            ).value = json.dumps(value)
+        return value, results
+
+    @classmethod
+    def prior_test_results(cls, _db, constructor_method=None, *args, **kwargs):
```

Comment:
Is this the method that I will use to display the results of the self tests from the db, or are you expecting the tests to re-run when an admin presses a "run tests" button?

Reply:
I don't know what it should look like, but I think you should always show the latest test results, with the button to re-run. The admin shouldn't have to explicitly request the results. Particularly if there's a failure, it should be prominently displayed up front.

Reply:
That's fine. The controller can call `prior_test_results` whenever it wants, even if no one has explicitly asked for the test results.

The only problem is that sometimes instantiating the object (which is necessary to call `external_integration` to retrieve the test results) triggers API requests. If the API configuration is broken, you might get a failure before you can even retrieve the test results. That's not a big problem because the controller can catch the exception and show that there's a problem.

However, a larger problem is that the API request might succeed but take a really long time (Odilo is currently in this situation). So now we're looking at a 30-second load time on the collections page even if all the tests technically pass.

The long-term solution is to make sure just instantiating the class doesn't make any network requests. I think we'd only have to change Odilo and Overdrive.